### PR TITLE
Force vpc endpoint type to gateway in govcloud partition.

### PR DESCRIPTION
**Pull request by jmcarp**
The aws govcloud partition only supports gateway type vpc endpoints
and doesn't return a type when listing connections. To fix, temporarily
force vpc endpoint types to gateway if partition is govcloud.

After vpc endpoint types were added in 1.9, govcloud vpc endpoints
want to destroy and recreate on every apply, since the aws api doesn't
return an endpoint type.

cc @wjwoodson

### CODE REVIEW

**Comment by jmcarp** [comment]
By the way, it might be useful to add an integration test that creates a resource, then verifies that the plan is empty.

File: aws/resource_aws_vpc_endpoint.go [hunk 1]
```diff
@@ -341,14 +341,18 @@ func vpcEndpointWaitUntilAvailable(d *schema.ResourceData, conn *ec2.EC2) error
...
 	d.Set("vpc_id", vpce.VpcId)
 
 	serviceName := aws.StringValue(vpce.ServiceName)
 	d.Set("service_name", serviceName)
-	vpceType := aws.StringValue(vpce.VpcEndpointType)
-	d.Set("vpc_endpoint_type", vpceType)
+	if isGovCloud {
+		d.Set("vpc_endpoint_type", ec2.VpcEndpointTypeGateway)
+	} else {
+		vpceType := aws.StringValue(vpce.VpcEndpointType)
```
Comment (ewbankkit) [inline aws/resource_aws_vpc_endpoint.go]: No need to call `aws.StringValue` on the VPC endpoint type as `d.Set` will handle `*string` parameters. I only call it above for the service name as the value is used later.

File: aws/resource_aws_vpc_endpoint.go [hunk 2]
```diff
@@ -341,14 +341,18 @@ func vpcEndpointWaitUntilAvailable(d *schema.ResourceData, conn *ec2.EC2) error
 	return nil
 }
 
-func vpcEndpointAttributes(d *schema.ResourceData, vpce *ec2.VpcEndpoint, conn *ec2.EC2) error {
+func vpcEndpointAttributes(d *schema.ResourceData, vpce *ec2.VpcEndpoint, conn *ec2.EC2, isGovCloud bool) error {
```
Comment (ewbankkit) [inline aws/resource_aws_vpc_endpoint.go]: Consider changing the function signature to
`func vpcEndpointAttributes(d *schema.ResourceData, meta interface{}, vpce *ec2.VpcEndpoint) error`
as both `conn` and `isGovCloud` can be derived from `meta`?

**Comment by jmcarp** [comment]
Updated. WDYT @ewbankkit ?

**Comment by jmcarp** [comment]
Pinging @radeksimko or @bflad for review when you have time.

**Comment by jmcarp** [comment]
ping @bflad

**Comment by bflad** [comment]
Can you provide a previously failing configuration and the outputs from plan/apply?

**Comment by jmcarp** [comment]
@bflad: terraform plan/apply doesn't fail because of this issue--it recreates the endpoint on every apply. Every `plan` says something like `module.stack.module.base.module.vpc.aws_vpc_endpoint.private-s3 (new resource required)` because it wants to change the endpoint type: `vpc_endpoint_type:          "" => "Gateway" (forces new resource)`. Every `apply` recreates the resource, but since govcloud ignores the endpoint type parameter, we repeat the cycle on the next `plan`.

**Comment by jmcarp** [comment]
@bflad: does that description make sense? Let me know if more output would be helpful.

File: aws/resource_aws_vpc_endpoint.go [hunk 3]
```diff
@@ -341,14 +341,17 @@ func vpcEndpointWaitUntilAvailable(d *schema.ResourceData, conn *ec2.EC2) error
 	return nil
 }
 
-func vpcEndpointAttributes(d *schema.ResourceData, vpce *ec2.VpcEndpoint, conn *ec2.EC2) error {
+func vpcEndpointAttributes(d *schema.ResourceData, vpce *ec2.VpcEndpoint, meta interface{}) error {
```
Comment (bflad) [inline aws/resource_aws_vpc_endpoint.go]: Let's revert this, I'll explain below. 😄

File: aws/resource_aws_vpc_endpoint.go [hunk 4]
```diff
@@ -341,14 +341,17 @@ func vpcEndpointWaitUntilAvailable(d *schema.ResourceData, conn *ec2.EC2) error
...
 	d.Set("state", vpce.State)
 	d.Set("vpc_id", vpce.VpcId)
 
 	serviceName := aws.StringValue(vpce.ServiceName)
 	d.Set("service_name", serviceName)
-	vpceType := aws.StringValue(vpce.VpcEndpointType)
-	d.Set("vpc_endpoint_type", vpceType)
+	if meta.(*AWSClient).IsGovCloud() && aws.StringValue(vpce.VpcEndpointType) == "" {
```
Comment (bflad) [inline aws/resource_aws_vpc_endpoint.go]: Instead of relying on a `meta.(*AWSClient).IsGovCloud()` check, which one day may become obsolete, let's just perform the second check `aws.StringValue(vpce.VpcEndpointType) == ""` with a comment above it why the logic is in place. 👍

**Comment by jmcarp** [comment]
Sounds good, updated.

**Comment by bflad** [comment]
This has been released in [version 1.12.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1120-march-23-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by lorengordon** [comment]
Thanks @bflad, tested and it does indeed solve the problem I was having.
